src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -820,6 +845,247 @@ CoordinatedRemoteTransactionsAbort(void)
 
 
 /*
+ * RemoteTransactionsSavepointBegin sends the SAVEPOINT command for the given
+ * sub-transaction id to all connections participating in the current transaction.
+ */
+void
+RemoteTransactionsSavepointBegin(SubTransactionId subId)
+{
+	dlist_iter iter;
+
+	/* asynchronously send SAVEPOINT */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		StartRemoteTransactionSavepointBegin(connection, subId);
+	}
+
+	/* and wait for the results */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		FinishRemoteTransactionSavepointBegin(connection, subId);
+
+		if (!transaction->transactionFailed)
+		{
+			transaction->lastUnfailedSubXact = subId;
+		}
+	}
+}
+
+
+/*
+ * RemoteTransactionsSavepointRelease sends the RELEASE SAVEPOINT command for the
+ * given sub-transaction id to all connections participating in the current transaction.
+ */
+void
+RemoteTransactionsSavepointRelease(SubTransactionId subId)
+{
+	dlist_iter iter;
+
+	/* asynchronously send SAVEPOINT */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		StartRemoteTransactionSavepointRelease(connection, subId);
+	}
+
+	/* and wait for the results */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		FinishRemoteTransactionSavepointRelease(connection, subId);
```

Comment:
Yes, this is ok. We are just interested in the last "SAVEPOINT" command that successfully executed before transaction failure. Whether that savepoint was released or not doesn't matter for us.

The key point to understanding this is that first sub-transaction always gets subId=2 by postgres, and n-th sub-transaction always gets subId=n+1. If our last successful sub-transaction is SAVEPOINT m, then rolling back to this sub-transaction or any sub-transaction with lesser id recovers the transaction from failure. But rolling back to sub-transactions with greater id doesn't recover the transaction because those SAVEPOINTs happened when we already had a failure.

Sure, there might be some of these savepoints that have been already released, but we don't need to worry about it since postgres already catches this before control reaches our callback in transaction_management.c.

So we don't need to keep an exact stack of what is active or released to handle rolling back.